addon/index.d.ts

```diff
@@ -0,0 +1,828 @@
+import ComputedProperty from '@ember/object/computed';
+
+export type TaskGenerator<T> = Generator<any, T, unknown>;
```

Comment:
```suggestion
export type TaskGenerator<T> = Generator<any, T, any>;
```

Should the third parameter (the LHS of `yield ...`, i.e. what it "returns") be `unknown` or `any`?

Due to TS limitations, there is no way we can type that properly. What the `yield` "returns" is based on what you yielded, and that is just not possible to type. `await` has a similar problem but is hardcoded to work in the TS compiler.

The first parameter (the RHS of the `yield ...`, i.e. what you are allowed to yield to ember-concurrency) is typed to `any` because there is no real utility to limit it, it really can be anything (including nothing – you may just be yielding so ember-concurrency can pause to check for `isDestroying` and such).

The third parameter is typed to `unknown`, but I am not sure if that is too strict. In practice, it will force every consumer to check the value. If you are in the mood for it you may write an assertion:

```ts
class Foo {
  @task *myTask() {
    let foo /* unknown */ = yield someStringPromise;
    assert('foo should have resolved into a string!', typeof foo === 'string');
    // now foo is a string
  }
}
```

...that's fine when it's easy enough to check, but it's this is really checking for bugs in the library than in your code, and you would never have written that assertion there if not for TS. For more complex objects you may end up needing to write elaborate type assertion functions that are otherwise not needed. In practice, a lot of people is just going to end up writing this anyway, every single time they yield in a task:

```ts
class Foo {
  @task *myTask() {
    let foo: MyType = (yield someStringPromise) as any;
  }
}
```

If that's what everyone is going to do at the end of the day, it just seems a little silly not to type it as `any` in the first place. In practice, so long as you know the type of the RHS, there is very little reason not to trust e-c to call you with the correct thing on the LHS. If you are worried that you would have mistyped the LHS, you could do this:

```ts
type Resolved<T> = T extends PromiseLike<infer R> ? R : T;

class Foo {
  @task *myTask() {
    let foo: Resolved<typeof someStringPromise> = yield someStringPromise;
  }
}
```

That seems like a pretty good compromise between correctness and verbosity, and led me to thinking we should go with `any` instead of `unknown` here. Thoughts?

Reply:
Ah, I realized my mistake (I actually thought about it when I typed this), in practice it doesn't *really* matter what we typed here, because the types are never really used for inference in the position I was refereeing to above. Since it's inside the body of a generator function, TypeScript is just going to do whatever it normally does to generator functions without ever considering the type we have here. This type is only used when assigning a generator function to `task(...)`, etc, as long as the type is wide enough that TS allows assigning any valid task function to positions where we _take_ one as a `TaskFunction<T>`, it doesn't really matter.

Reply:
I think I'm on the same page given the constraints TS gives us. I would recommend we *very* carefully document this, however, because it's the kind of thing that could easily bite people by letting `any` leak in when they're unaware of it.

Reply:
I was wrong _again_. Here is how it comes up in practice (typing the return type of the generator function) and the ergonomic differences:

[Playground](https://www.typescriptlang.org/play/index.html?experimentalDecorators=true&ssl=30&ssc=1&pln=31&pc=1#code/GYVwdgxgLglg9mABAQwM6oKYCcoAoIIAmMsCAXCmAJ4CUFamOqiBYxpSA3ogL4BQoSB0RQ0Aa1wA6aciwBzVPWoBtALp1KVRN358oVAA4ZEAZShYY0ACriA4hjDZkUOFgA8VgHyIAvInuOWM6ubsjUADSIVpHgYmBwAO5gngDceobGADJwcJg2qGIBTi7uXr7+DsUhYVSR0ZqpfOlGiABKGKhwADYAbhiEHt5+VogYAB5QDoTMAApYcAC2MJiZMGIYbjBgwNht3gD8bYgUVmnNxu1QIFhgW3KD5VIy8oqaajS+3qfnbR3dfYQrBkPKMJlNmJdrrcwPcap4hlFQZM2BCMFcbnc3HNFssMKt1pttrtWvDEIdWsdfuiwECjIMUogmhAugxEAAxHLaPiIHmIAACogKiAAVKhzJYoPkJBozBZrHZKkESm4xRYYd5ONzedqumjEAlEuVsUtMJIsH9ehhcAByA0Ja00NLanV682dS2EcpUGAYLqeu1O508hjYPD6IxwYCIN3-fq+Hx+a2qu4Os5Bnm6qCIMUgYBRvze32eqAAC2Wkhzedwjq1QZDOFw4YwkezVzz8cTyZhqaa6ejaKhrdzwEkLgAqgYjFgAMJoK0fADU-fdANHcAnU9nmGrgZ5um1AvEIq6OTy4mrFGyuQwUqKSpCXbkGtrzsz+sNfmNuLNFr6NrtPZ9m+MYehQ7Qrv0bhNi2doIoWfrvgkabpm+lbAGBv79LSGylssyhJm2wDWqocE+ghuGoBWhE7r26bmtSQ55muG7YFu86IEuIGruOk6sXONHavuvJoFQkCMcAF6IF+mAquK6pcn29GDgRw7WruvB8PwQA)

This doesn't look so bad since the types are just strings and the assertion is easy to write, but basically see my first comment for the tradeoffs.

Reply:
Talked to @chriskrycho on [Discord](https://discordapp.com/channels/480462759797063690/484421406659182603/717868405587312682), it seems like we are leaning towards using `any` here and documenting/recommending the `Resolved` type trick

Reply:
Notably, both the `Resolved` type trick *and* the `assert` approach will work if we type as `any` here: `assert` will correctly narrow both `unknown` and `any`.

Reply:
Right, the main question is do we want to _force_ the consumer to do a check.

Reply:
As you noted, it's a *very* frustrating experience to *have* to when the type *is* actually knowable and local. I don't have a strong opinion here (and I normally *very* strongly favor `unknown` and forcing the check).

cc. also @dfreeman @jamescdavis

Reply:
Maybe I'll just go ahead and write a compiler plugin that transforms `@task async` methods into generators and no have to worry about this 😄

Reply:
Anddd done: https://github.com/chancancode/ember-concurrency-async

Reply:
I went with `any` in the end.